### ISSUE TITLE
Memory leak when using request.proxy()

### ISSUE THREAD

**Issue by heyphets** [drachtio/drachtio-srf#102]
Hi Dave, 

It appears I've come across a memory leak in drachtio-srf when using request.proxy() method. The memory consumed by application constantly increasing linearly with number of calls placed. In a Chrome DevTools profile I see an increasing number of streams in the memory, the strings appear to be full body of the incoming messages. At some point (at about 1G heap size in my environment) GC appears to run which spikes the CPU and affects the call processing. 

To replicate I'm using the following code:
```javascript
const Srf = require('drachtio-srf');
const srf = new Srf();


srf.connect({
  'host': '127.0.0.1',
  'port': 9022,
  'secret': 'cymru'
});

srf.on('connect', (err, hp) => {
  if (err) throw err;
  console.log(`connected to drachtio listening on ${hp}`);
});
srf.on('error', (err) => console.error(err));


srf.invite(async(req, res) => {
  try {
    const targets = ['sip:10.8.56.109:5060'];
    req.proxy({
      destination: targets,
      recordRoute: false,
      provisionalTimeout: '2s',
      finalTimeout: '4s'
    });
  }
  catch (err) {
    console.error({err}, 'Error proxying invite');
  }
});
````
where 10.8.56.109 is machine hosting SIPp. 

The SIPp script answering the call at 10.8.56.109:5060: 
```xml
<?xml version="1.0" encoding="ISO-8859-1" ?>
<!DOCTYPE scenario SYSTEM "sipp.dtd">

<!-- This program is free software; you can redistribute it and/or      -->
<!-- modify it under the terms of the GNU General Public License as     -->
<!-- published by the Free Software Foundation; either version 2 of the -->
<!-- License, or (at your option) any later version.                    -->
<!--                                                                    -->
<!-- This program is distributed in the hope that it will be useful,    -->
<!-- but WITHOUT ANY WARRANTY; without even the implied warranty of     -->
<!-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      -->
<!-- GNU General Public License for more details.                       -->
<!--                                                                    -->
<!-- You should have received a copy of the GNU General Public License  -->
<!-- along with this program; if not, write to the                      -->
<!-- Free Software Foundation, Inc.,                                    -->
<!-- 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA             -->
<!--                                                                    -->
<!--                 Sipp default 'uas' scenario.                       -->
<!--                                                                    -->

<scenario name="Basic UAS responder">
  <!-- By adding rrs="true" (Record Route Sets), the route sets         -->
  <!-- are saved and used for following messages sent. Useful to test   -->
  <!-- against stateful SIP proxies/B2BUAs.                             -->
  <!-- Adding ignoresdp="true" here would ignore the SDP data: that     -->
  <!-- can be useful if you want to reject reINVITEs and keep the       -->
  <!-- media stream flowing.                                            -->
  <recv request="INVITE" crlf="true">
  </recv>

  <!-- The '[last_*]' keyword is replaced automatically by the          -->
  <!-- specified header if it was present in the last message received  -->
  <!-- (except if it was a retransmission). If the header was not       -->
  <!-- present or if no message has been received, the '[last_*]'       -->
  <!-- keyword is discarded, and all bytes until the end of the line    -->
  <!-- are also discarded.                                              -->
  <!--                                                                  -->
  <!-- If the specified header was present several times in the         -->
  <!-- message, all occurrences are concatenated (CRLF separated)       -->
  <!-- to be used in place of the '[last_*]' keyword.                   -->

  <send>
    <![CDATA[

      SIP/2.0 180 Ringing
      [last_Via:]
      [last_From:]
      [last_To:];tag=[pid]SIPpTag01[call_number]
      [last_Call-ID:]
      [last_CSeq:]
      Contact: <sip:[local_ip]:[local_port];transport=[transport]>
      Content-Length: 0

    ]]>
  </send>

  <send>
    <![CDATA[

      SIP/2.0 200 OK
      [last_Via:]
      [last_From:]
      [last_To:];tag=[pid]SIPpTag01[call_number]
      [last_Call-ID:]
      [last_CSeq:]
      Contact: <sip:[local_ip]:[local_port];transport=[transport]>
      Content-Type: application/sdp
      Content-Length: [len]

      v=0
      o=user1 53655765 2353687637 IN IP[local_ip_type] [local_ip]
      s=-
      c=IN IP[media_ip_type] [media_ip]
      t=0 0
      m=audio [media_port] RTP/AVP 0
      a=rtpmap:0 PCMU/8000

    ]]>
  </send>


  <!-- Keep the call open for a while in case the 200 is lost to be     -->
  <!-- able to retransmit it if we receive the BYE again.               -->
  <timewait milliseconds="4000"/>


  <!-- definition of the response time repartition table (unit is ms)   -->
  <ResponseTimeRepartition value="10, 20, 30, 40, 50, 100, 150, 200"/>

  <!-- definition of the call length repartition table (unit is ms)     -->
  <CallLengthRepartition value="10, 50, 100, 500, 1000, 5000, 10000"/>

</scenario>
```

The SIPp script generating calls: 
```xml
<?xml version="1.0" encoding="ISO-8859-1" ?>
<!DOCTYPE scenario SYSTEM "sipp.dtd">

<!-- This program is free software; you can redistribute it and/or      -->
<!-- modify it under the terms of the GNU General Public License as     -->
<!-- published by the Free Software Foundation; either version 2 of the -->
<!-- License, or (at your option) any later version.                    -->
<!--                                                                    -->
<!-- This program is distributed in the hope that it will be useful,    -->
<!-- but WITHOUT ANY WARRANTY; without even the implied warranty of     -->
<!-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the      -->
<!-- GNU General Public License for more details.                       -->
<!--                                                                    -->
<!-- You should have received a copy of the GNU General Public License  -->
<!-- along with this program; if not, write to the                      -->
<!-- Free Software Foundation, Inc.,                                    -->
<!-- 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA             -->
<!--                                                                    -->
<!--                 Sipp 'uac' scenario with pcap (rtp) play           -->
<!--                                                                    -->

<scenario name="UAC with media">
  <!-- In client mode (sipp placing calls), the Call-ID MUST be         -->
  <!-- generated by sipp. To do so, use [call_id] keyword.                -->
  <send retrans="500">
    <![CDATA[

      INVITE sip:[remote_ip]:[remote_port] SIP/2.0
      Via: SIP/2.0/[transport] [local_ip]:[local_port];branch=[branch]
      From: <sip:sipp@[local_ip]:[local_port]>;tag=[pid]SIPpTag09[call_number]
      To: <sip:[service]@[remote_ip]:[remote_port]>
      Call-ID: [call_id]
      CSeq: 1 INVITE
      Contact: sip:sipp@[local_ip]:[local_port]
      Max-Forwards: 70
      Require: siprec
      MIME-Version: 1.0
      Content-Type: multipart/mixed;boundary=unique-boundary-1
      Content-Length: [len]

      --unique-boundary-1
      Content-Type: application/sdp

      v=0
      o=- 20628849 767165 IN IP4 10.6.8.55
      s=-
      c=IN IP4 10.6.8.55
      t=0 0
      m=audio 53298 RTP/AVP 18 0 8 101
      a=rtpmap:18 G729/8000
      a=fmtp:18 annexb=no
      a=rtpmap:0 PCMU/8000
      a=rtpmap:8 PCMA/8000
      a=rtpmap:101 telephone-event/8000
      a=fmtp:101 0-15
      a=maxptime:20
      a=label:76965361
      a=sendonly
      m=audio 53302 RTP/AVP 0 101
      c=IN IP4 10.6.8.55
      a=rtpmap:0 PCMU/8000
      a=rtpmap:101 telephone-event/8000
      a=fmtp:101 0-16
      a=label:76965362
      a=sendonly

      --unique-boundary-1
      Content-Type: application/rs-metadata+xml
      Content-Disposition: recording-session

      <?xml version='1.0' encoding='UTF-8'?>
      <recording xmlns='urn:ietf:params:xml:ns:recording'>
      	<datamode>complete</datamode>
      	<session id="/FEXR15pQadPigc46TF5kQ==">
      		<associate-time>2020-04-22T10:10:11</associate-time>
      		<extensiondata xmlns:apkt="http:/acmepacket.com/siprec/extensiondata">
      			<apkt:ucid>3194409394-0032146609-2832138240-0001197432</apkt:ucid>
      			<apkt:callerOrig>true</apkt:callerOrig>
      		</extensiondata>
      	</session>
      	<participant id="sC56dHKjSi9q6Fxq5RlmKw==" session="/FEXR15pQadPigc46TF5kQ==">
      		<nameID aor="sip:+19999999999@2.2.2.2">
      			<name>&quot;+19999999999&quot;</name>
      		</nameID>
      		<send>B/MLI0o7Rnxo88URCUsiCQ==</send>
      		<associate-time>2020-04-22T10:10:11</associate-time>
      		<extensiondata xmlns:apkt="http://acmepacket.com/siprec/extensiondata">
      			<apkt:callingParty>true</apkt:callingParty>
      		</extensiondata>
      	</participant>
      	<participant id="4NwcRvRaTDJfSHW2CGD7pQ==" session="/FEXR15pQadPigc46TF5kQ==">
      		<nameID aor="sip:18440000000@1.1.1.1">
      			<name>18443271137</name>
      		</nameID>
      		<send>pUavJ5WlS8RvLOsMu0uVsQ==</send>
      		<associate-time>2020-04-22T10:10:11</associate-time>
      		<extensiondata xmlns:apkt="http://acmepacket.com/siprec/extensiondata">
      			<apkt:callingParty>false</apkt:callingParty>
      		</extensiondata>
      	</participant>
      	<stream id="B/MLI0o7Rnxo88URCUsiCQ==" session="/FEXR15pQadPigc46TF5kQ==">
      		<label>76965361</label>
      		<mode>separate</mode>
      		<associate-time>2020-04-22T10:10:11</associate-time>
      	</stream>
      	<stream id="pUavJ5WlS8RvLOsMu0uVsQ==" session="/FEXR15pQadPigc46TF5kQ==">
      		<label>76965362</label>
      		<mode>separate</mode>
      		<associate-time>2020-04-22T10:10:11</associate-time>
      	</stream>
      </recording>
      --unique-boundary-1--
    ]]>
  </send>

  <recv response="100" optional="true">
  </recv>

  <recv response="180" optional="true">
  </recv>

  <!-- By adding rrs="true" (Record Route Sets), the route sets         -->
  <!-- are saved and used for following messages sent. Useful to test   -->
  <!-- against stateful SIP proxies/B2BUAs.                             -->
  <recv response="200">
  </recv>



  <!-- definition of the response time repartition table (unit is ms)   -->
  <ResponseTimeRepartition value="10, 20, 30, 40, 50, 100, 150, 200"/>

  <!-- definition of the call length repartition table (unit is ms)     -->
  <CallLengthRepartition value="10, 50, 100, 500, 1000, 5000, 10000"/>

</scenario>
```

**Comment by davehorton** [comment]
Good catch.  There were some retained objects that were not getting freed in the proxy scenario.  Can you test the fix that is in drachtio-srf@4.4.47 ?

**Comment by heyphets** [comment]
Thanks Dave, this worked. I've run 50000 calls, memory didn't grow beyond some 50Mb.
